citro3d/src/uniform.rs

```diff
@@ -1,15 +1,17 @@
 //! Common definitions for binding uniforms to shaders. This is primarily
 //! done by implementing the [`Uniform`] trait for a given type.
 
-use crate::math::Matrix;
+use std::ops::Range;
+
+use crate::math::{FVec4, IVec, Matrix4};
 use crate::{shader, Instance};
 
 /// The index of a uniform within a [`shader::Program`].
-#[derive(Copy, Clone, Debug)]
-pub struct Index(i8);
+#[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord)]
+pub struct Index(u8);
 
-impl From<i8> for Index {
-    fn from(value: i8) -> Self {
+impl From<u8> for Index {
```

Comment:
I'm fine with changing the representation to `u8` here, but I guess maybe we should implement `TryFrom` to make sure the index is always in the range `0..=i8::MAX`? Not sure what the behavior would be if you passed a negative value or something to the uniform APIs...

Reply:
you can't because it'd be caught by the checks in `Uniform::bind`, its `u8` cause that makes more sense for an index, `<0` isn't valid for them and is only returned by the functions in the C API so it can report error

Reply:
I guess I'm wondering what would happen if someone did `let i = Index::from(200)` and then tried to use that index later on in a `bind` call... Maybe we should remove this `From` impl entirely and make it so the only way to get an `Index` is via `Program::get_uniform`.

Reply:
> I guess I'm wondering what would happen if someone did `let i = Index::from(200)` and then tried to use that index later on in a `bind` call... Maybe we should remove this `From` impl entirely and make it so the only way to get an `Index` is via `Program::get_uniform`.

yeah it'd panic on the bind call

Reply:
actually I think I can make this even safer by pulling the valid register range out of the shader program in get uniform and putting it in index. this would let us assert and prevent binding mismatching uniforms (e.g. trying to bind a 4x4 to a 1x4). should I add that?

Reply:
That sounds great! The more safe we can make the APIs the better, in my opinion

Reply:
yeah tried this uh something is going wrong with parsing the shader file but it seems to work weirdly and I can't work it out so I'm going to leave it for now, can always be done later